### PR TITLE
Update Tern & Acorn

### LINKED ISSUE #11568: JS Code Hints won't work when there's ".." in a HTML file
1. Take the HTML below and paste it into a HTML file
2. In line 4, write out `test.`
3. Remove the two trailing dots and save
4. Restart Brackets
5. Repeat step 2

``` html
..
<script>
    var test = 1;
    
</script>

```

Result:
In step 2, you don't get JS Code Hints, whereas in step 5, you get Code Hints

### PR DESCRIPTION
**Pull request by marcelgerber**
Since Acorn's build process changed (there are no more prebuilt `dist` files in the repo, but in the npm package), I've switched from having it available as a submodule to just having the files in the file tree. Hope that's fine.
The build is based on my fork https://github.com/MarcelGerber/acorn, where I removed a build step that doesn't quite work with Brackets.

Fixes #11568, and hopefully some others, too.

Commit logs:
[Tern](https://github.com/marijnh/tern/compare/7606a6448a8f7a2aacd50d10d9752440689803e8...34549901036cb2c891dd8d08cdaefefce7f5d542) + [Release notes](http://ternjs.net/#releases)
[Acorn](https://github.com/marijnh/acorn/compare/78e1d7ada684868d92d09555408ae2df1812b5ae...0072255bc35d51062a6e92692166dc3f9244c5a3)

### CODE REVIEW

**Comment by MarcelGerber** [comment]
All tests pass now.

**Comment by MarcelGerber** [comment]
Woah. Looks like this upgrade actually finally fixes the problems we had with Ionic/Cordova apps (cc @MiguelCastillo - would be very kind of you if you could test that). That'd be absolutely awesome.

**Comment by ficristo** [comment]
Out of curiosity, can I know why they can't be dependencies in package.json?

**Comment by MarcelGerber** [comment]
The two libraries have to be located in the extension's folder, thus they cannot be part of the root `package.json`.
I don't know for sure why they are not part of a package.json in the `thirdparty` folder itself, but I guess it's just so you can start "hacking" Brackets in as few steps as possible, where `npm install` is not needed.

**Comment by MarcelGerber** [comment]
I just now updated this to feature the latest Acorn & Tern versions, where Tern now has ES6 support (#11644).
If you'd rather have the "normal" Tern upgrade first (it's presumably more stable), let me know and I'll factor the ES6 one out into a new PR.

cc @redmunds @nethip for review

**Comment by MiguelCastillo** [comment]
@MarcelGerber - tern 0.15 is just released a few hours ago.  I think that it will be worth doing that version since ES6 has been more thoroughly tested and bugs fixed.  That should definitely improve things.

Some things to keep in mind is that there is a new plugin called `module`, which needs to be enabled in order to es6 to work well. For example, this is what my tern config looks like now.

```
{
	"libs": [
		"browser",
		"chai",
		"ecma5",
		"ecma6",
		"browser",
		"jquery"
	],
	"loadEagerly": [],
	"async": true,
	"plugins": {
		"doc_comment": {},
		"requirejs": {},
		"es_modules": {},
		"modules": {}
	}
}
```

**Comment by MarcelGerber** [comment]
Thanks @MiguelCastillo 
Updating Tern was already on my agenda, and I just committed and pushed that upgrade.

I'll try to look at the `modules` plugin after my 4-day-trip to Berlin, on Sunday/Monday.
Or if you feel like it, feel free to do this yourself :)

**Comment by abose** [comment]
As this is a core change, Moving to next release as all external integrations happen at the beginning of the cycle.
@MarcelGerber Will target this immediately after rel 1.5 is out.

**Comment by ingorichter** [comment]
@abose what is the timeline for the 1.5 release? I'm just asking, since it's not clear to me, if this means we are holding off PR's for days, weeks or months.

**Comment by abose** [comment]
1.5 is due within the next 2 weeks and the move to release branch is by next week.
So the holdoff is probably till next week :)

**Comment by ingorichter** [comment]
That's great @abose! Then let's hold it off until then.

**Comment by MarcelGerber** [comment]
Updated once again to include the latest & greatest.

**Comment by busykai** [comment]
@MarcelGerber, accidentally tried more or less the same and hit the unfortunate change in `acorn` as well. I think replacing it with a copy is the right thing to do, except these doubts:

1. why do we need all the rest of the files in `thirdparty/acorn`? perhaps we only need to keep the LICENSE.md, remove the rest except `dist/*`, and move all the files from `dist/` to the root (or leave it there if `tern` relies on them being in `dist`).

2. we need to be explicit in the comments which version/commit id of acorn is being committed. since it's not a submodule any longer, it's needed for the bookkeeping.

**Comment by MarcelGerber** [comment]
@busykai Where'd you suggest putting the commit SHA (we already have the version embedded in code)?
Otherwise, we could also simply settle with release versions, so we can rely on the version number by itself, not having to worry about the SHA.

**Comment by busykai** [comment]
@MarcelGerber, I was thinking additionally a version tag/sha in the commit comment. So that you could see with `git log` the history of the versions.

**Comment by MarcelGerber** [comment]
Ah yes, that makes sense.
But considering I'll rebase these commits away anyhow at some point, we can do that later on.
For now, I'd like to see whether this actually works as expected.

**Comment by busykai** [comment]
@MarcelGerber, understood. What about removing all the acorn junk? We don't need it there (provided we know where to take it from): since it's no longer a submodule, having it in `thirdparty/acorn` does not make any sense.

I'll take a look at it functionally and run some tests this weekend. @MiguelCastillo also was going to get back to this.

**Comment by MiguelCastillo** [comment]
@MarcelGerber @busykai Ok, I had to read the whole thing again. :)

I'd really prefer using npm packages.  It just makes our lives easier because we don't need to worry maintaining yet another repo.

One thing that I think we can do as a separate PR is to move thirdparty deps into the package.json as way to manage dependency upgrades. And we can manage copying these modules from node_modules into thirdparty (or any other place) in our gruntfile.js, which we already do for a few other things. I can look into that separately.  But for now, adding tern as a dependency and adding a config in the gruntfile.js to copy the needed files to the right place will be a more straight forward workflow for future upgrades.  I see that there are concerns with setting up brackets for hacking.  I am not too concerned about that because we will be checking in the dependencies into the thirdparty directory anyways.  Upgrades will continue to be a manual process of running `npm install` and putting out a PR to merge the new stuff into the correct place. No more `git submodule init -u ... -whtver` please :D

I am going to pull this down right now and take it out for test.

**Comment by ficristo** [comment]
@MiguelCastillo you should take a look at https://github.com/adobe/brackets/issues/12006

**Comment by MiguelCastillo** [comment]
@ficristo Well thank you very much sir!

**Comment by petetnt** [comment]
RE: the requirement for the files being under `/dist`

You can map the paths in https://github.com/adobe/brackets/blob/master/src/main.js#L31 to your liking; you just need to make sure that `acorn_loose` which depends on `acorn` (iirc) can find it. 

See the `react` example that got merged some time last week https://github.com/petetnt/brackets/blob/71d1442c083d48d671ac3905ed70ee2ac37bebe3/src/main.js#L31 (`ReactDOM` requires `react`).

**Comment by MarcelGerber** [comment]
@MiguelCastillo @busykai @abose I've once again updated to the latest versions of both Acorn and Tern. I haven't done much testing, but it seems to at least work.

Also, I've stripped out most of the Acorn files, so we're only left with the `dist` folder plus AUTHORS, README and LICENSE files.

Would love if you could take this for a test (and I will too!)

**Comment by MiguelCastillo** [comment]
Ok great! I am pulling it right now.

**Comment by MiguelCastillo** [comment]
@MarcelGerber help me understand where we stand with this PR.  We are continuing to use tern as a submodule and we have pulled out acorn to be from the npm repo?

**Comment by MarcelGerber** [comment]
Tern is still a submodule, that's true.

Acorn is based of a build of my fork (https://github.com/MarcelGerber/acorn).
We cannot use a submodule anymore because over in the repo, only source files are available, no built ones.
But we also cannot use the npm version because there's some strange custom require logic going on (apparently to enable usage with Browserify), which I couldn't get to work with our code (@busykai didn't either).
So I went with a custom build based off a fork for now (which has the custom require removed: https://github.com/MarcelGerber/acorn/commit/77e0fc99d47bcc4aff8e978a291fa3d9de635a1c). Not the best solution, I know that.

Also, I've just pushed a new commit to workaround this issue: https://github.com/ternjs/acorn/issues/412

**Comment by ficristo** [comment]
IMHO we should land https://github.com/adobe/brackets/pull/11948 and then use the node counterpart.

**Comment by MarcelGerber** [comment]
Updated again.
Works fine and all tests pass.

As Acorn updated their build process, we... sadly still cannot use its output as is.
So I applied another patch, which you can see here: https://github.com/MarcelGerber/acorn/tree/brackets-fix

**Comment by ficristo** [comment]
I really want this for next release, I only think #11948 is a better approach in the long term.

**Comment by zaggino** [comment]
Guys, can we please stick to npm dependencies or submodules and not push heaps of thirdparty code to the repository itself?

**Comment by MarcelGerber** [comment]
@ficristo @zaggino I just wanted to get it to the latest version so we don't get lost in too old a version.
If we can land both of these other PRs before this one, that's all fine by me. This is more or less a proof-of-work, and a proof that it still works :)

**Comment by MiguelCastillo** [comment]
Unfortunately not using npm is the primary reason I have been reluctant to merge this. @marcelgerber, do you need help transitioning this over to be an npm dependecy?

**Comment by MarcelGerber** [comment]
@MiguelCastillo Is there any reference PR or something along these lines?
Otherwise, if you like, you can certainly help me on that. Feel free to just push to this branch :)

**Comment by MiguelCastillo** [comment]
You mean my preference for npm? Yeah I added a comment earlier on in this thread.  Perhaps I didn't communicate it as well as I could have though.  :)  Sorry about that.  Just for reference https://github.com/adobe/brackets/pull/11569#issuecomment-186615827

But I will carve out some time today to make progress on that migration.  I believe I had done it before and it was not too bad.  The one thing I remember was a lil bit tricky is that the JS autocomplete that uses tern is an extension... So the grunt-task had to copy the files there, which is not a big deal.  I am just remembering what I had to do that.

**Comment by MiguelCastillo** [comment]
@zaggino can you clarify what you mean by "not push heaps of thirdparty code to the repository itself?"  Particularly the repository...  Because we do check stuff into the brackets src/thirdparty as I am sure you know.  I just wanna make sure we are on the same page for the migration.

**Comment by zaggino** [comment]
@MiguelCastillo Unless the stuff in `src/thirdparty` is
a) not available in npm
b) modified for brackets from the original npm version
there's no reason it shouldn't be a npm dependency installed by npm install - the practice of checking in 10,000 lines of code when real changes are a few lines and updated dependency version is simply bad and I'll repeat that in every PR I see doing that.

**Comment by MiguelCastillo** [comment]
@zaggino Yeah no disagreements there regarding NPM.  I just wasnt clear what you meant by "and not push heaps of thirdparty code to the repository itself".  That could have been read as the brackets repository, in case we do add heaps of thirdparty code to the thirdparty directory.

We are on the same page.

**Comment by ficristo** [comment]
I raised the priority here: there are quite a few issue that could be fixed by this.

**Comment by zaggino** [comment]
As a first thing, can we add `acorn` to `package.json`'s `dependencies` and remove it from this PR?

**Comment by MarcelGerber** [comment]
@zaggino Feel free to go ahead. I'm off till Sunday and while I can apply small changes, I cannot do so for substantial changes.

**Comment by zaggino** [comment]
I can have a look once https://github.com/adobe/brackets/pull/12647 gets merged.

**Comment by zaggino** [comment]
@MarcelGerber done a few commits and rebased on master, have a look at the diff now.
Run extension tests - no issues. I think this is now good to review and merge.

**Comment by petetnt** [comment]
Re: https://github.com/adobe/brackets/pull/11569/commits/50cb0007c0f24c0d6f52852847604fe91d97926b

Cannot this be fixed in the `requirejs.config`? I remember doing it when I did some kind of fix for the `npm dependencies only` PoC.

**Comment by zaggino** [comment]
@petetnt I'm afraid not https://github.com/requirejs/requirejs/issues/1560

**Comment by zaggino** [comment]
Of course you're welcome to try, I spent too much time on in to try again :-)

**Comment by MarcelGerber** [comment]
After having a quick look, this looks good to me, but I couldn't test it.
Fee free to squash away my multitude of "Update" commits, we definitely don't need them anymore.

**Comment by zaggino** [comment]
Rebased, squashed & ready to review.

**Comment by MarcelGerber** [comment]
Hm, there's a merge conflict in `npm-install.js` now. As you definitely know that code better than I do, I'll leave it to you, ok?

**Comment by zaggino** [comment]
Resolved the conflict @MarcelGerber

**Comment by MarcelGerber** [comment]
Another failure: Use the global "use strict" for fix-acorn.js

**Comment by zaggino** [comment]
:facepalm: fixed

**Comment by ficristo** [comment]
If using node_modules or requirejs is too hard maybe we could go with browserify?
Just throwing the idea.

**Comment by zaggino** [comment]
a) introducing a new tool when you can fix the script with 6 lines of code doesn't appeal to me
b) this should be ported to NodeDomain for improved performance and then you won't have to fix acorn for requirejs

File: src/extensions/default/JavaScriptCodeHints/npm-shrinkwrap.json [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+    "acorn": {
+      "version": "3.3.0",
+      "from": "acorn@3.3.0",
+      "resolved": "https://registry.npmjs.org/acorn/-/acorn-3.3.0.tgz"
+    },
+    "buffer-shims": {
+      "version": "1.0.0",
+      "from": "buffer-shims@>=1.0.0 <2.0.0",
+      "resolved": "https://registry.npmjs.org/buffer-shims/-/buffer-shims-1.0.0.tgz"
+    },
```
Comment (MarcelGerber) [inline src/extensions/default/JavaScriptCodeHints/npm-shrinkwrap.json]: Is this file supposed to have all these packages in there?

**Comment by MarcelGerber** [comment]
LGTM, just keen to know if `npm-shrinkwrap.json` is supposed to contain all these modules.

File: src/extensions/default/JavaScriptCodeHints/npm-shrinkwrap.json (hunk 1 above)
Comment (zaggino) [inline src/extensions/default/JavaScriptCodeHints/npm-shrinkwrap.json]: yes - that's how shrinkwrap works, it's generated by `npm shrinkwrap` command, I'd write more about it but it's easy to google

**Comment by zaggino** [comment]
LGTM +1

**Comment by MarcelGerber** [comment]
Thank you @zaggino for the work you have done these last few days to bring this up to par and make it mergeable!

**Comment by ficristo** [comment]
I've run the `Extensions \ JavaScript Code Hinting` on Windows, OSX and Linux.
While on Windows everything seems smooth, on the other platforms the tests fails.
Could you check?
(There are other tests that do not pass, like JSQuickEdit but I think the problem is the same)

**Comment by zaggino** [comment]
PR here https://github.com/adobe/brackets/pull/12732

**Pull request merged**